fhttp/http_utils.go

```diff
@@ -207,131 +208,127 @@ func removeTrailingPercent(s string) string {
 	return s
 }
 
-// generateStatus from string, format: status="503" for 100% 503s
-// status="503:20,404:10,403:0.5" for 20% 503s, 10% 404s, 0.5% 403s 69.5% 200s
-func generateStatus(status string) int {
-	lst := strings.Split(status, ",")
-	log.Debugf("Parsing status %s -> %v", status, lst)
+// parse input, i.e. "XX:20", using parseFunc
+func parseEntry(entry, about, input string, parseFunc func(string) (int64, error)) (int64, float64, error) {
+	l2 := strings.Split(entry, ":")
+	if len(l2) != 2 {
+		log.Warnf("Should have exactly 1 : in %s list %s -> %v", about, input, entry)
+		return -1, -1, errors.New("invalid format")
+	}
+	s, err := parseFunc(l2[0])
+	if err != nil {
+		log.Warnf("Bad input %s %v -> %v, not a number before colon", input, about, l2[0])
+		return -1, -1, errors.New("invalid format")
+	}
+	percStr := removeTrailingPercent(l2[1])
+	p, err := strconv.ParseFloat(percStr, 32)
+	if err != nil || p < 0 || p > 100 {
+		log.Warnf("Percentage is not a [0. - 100.] number in %v -> %v : %v %f", input, percStr, err, p)
+		return -1, -1, errors.New("invalid format")
+	}
+	return s, p, nil
+}
+
+func getWeightedValue(input string, weights []float32, values []int64) int64 {
+	res := 100. * rand.Float32()
+	for i, v := range weights {
+		if res <= v {
+			if i >= len(values) {
+				return -1
+			}
+			log.Debugf("[0. - 100.] for %s roll %f got #%d -> %d", input, res, i, values[i])
+			return values[i]
+		}
+	}
+	log.Debugf("[0. - 100.] for %s roll %f no hit, return default value", input, res)
+	return -1 // default/reminder of probability table
```

Comment:
this could be used for types other than int (the http status code being an example even if it maps fine to an int), and for cases where -1 is a valid value maybe nil would be a better return for this case
ps: for status the default is ok not -1 and for delay it was 0 but I guess if you document (the function `getWeightedValue` is missing comments/doc even for internal it's useful) clearly maybe it's ok as is (but moves the testing for -1 in the callers)